Fix exception on fmap

caioaao opened the pull request:
Previously the added testcase would __throw__:

```clojure
java.lang.ClassCastException: cats.monad.exception.Failure cannot be cast to java.lang.Number
```

sovelten commented:
good catch!
can you add changelog and update version?

dchelimsky commented on src/state_flow/state.clj:
```diff
@@ -21,8 +20,17 @@
-                       [(f v) ns]))
+                     (let [mp ((e/wrap (p/-extract fv)) s)]
+                       (cond
+                         (e/failure? mp)
```
Only one of these cases is tested. Are both actually necessary?

caioaao commented:
Turns out the issue was more widespread. I've added testcases to cover all issues I found and refactored the error state monad to handle them appropriately.

caioaao commented:
tests are green :heavy_check_mark:

dchelimsky commented on test/state_flow/state_test.clj:
```diff
@@ -11,7 +10,24 @@
     (is (state/gets inc))
```
These can be removed now that all of them are being used in tests below.

sovelten commented:
What's the benefit of using a new Record to encapsulate state? Is it for being able to distinguish from the vanilla implementation from cats?

caioaao commented:
@sovelten it's to implement the correct `p/-extract`
